### PR TITLE
RecreateVirtualSpaceRegionsForDisplay when display is added

### PR DESCRIPTION
**Pull request by kylef**
In some circumstances when you add a display, and since some windows will be moved to it as they was previously on a removed display. The positions of these windows will need to be updated to suit the new display. This can happen when you unplug a display, and then make another available. For example, if your MacBook(Pro) is closed, and then you disconnect an external display and then open the display. Windows would remain sized for the layout in the previous display.

To reproduce use the following for a laptop, I am not sure if you can reproduce with an iMac but perhaps similar with MacPro.

- Plug in an external display to a laptop, close the laptop lid so there is effectively one display
- Unplug external display
- Open the laptop lid so the new display becomes active.

The above steps output the following debug log (from master):

```
2018-08-15 18:37:24 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayChanged'
2018-08-15 18:37:25 DEBUG: 1:desktop   21302:0:iTerm2:./bin/chunkwm  /Users/kyle/Projects/tmp/chunkwm
2018-08-15 18:37:25 PROFILE: #680:plugin.mm:GetAllVisibleWindowsForSpace = 0.12400000ms
2018-08-15 18:37:25 DEBUG: chunkwm:Tiling:focused_desktop_mode
2018-08-15 18:37:25 DEBUG: chunkwm:Tiling:focused_window_float
2018-08-15 18:37:25 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayRemoved'
2018-08-15 18:37:25 DEBUG: 459132913: display removed
2018-08-15 18:37:25 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayAdded'
2018-08-15 18:37:25 DEBUG: 1104977153: display added
2018-08-15 18:37:25 DEBUG: chunkwm: processing event of type 'ChunkWM_PluginBroadcast'
2018-08-15 18:37:25 DEBUG: chunkwm: processing event of type 'ChunkWM_PluginBroadcast'
2018-08-15 18:37:32 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayChanged'
2018-08-15 18:37:33 DEBUG: 1:desktop   21302:0:iTerm2:./bin/chunkwm  /Users/kyle/Projects/tmp/chunkwm
2018-08-15 18:37:33 PROFILE: #680:plugin.mm:GetAllVisibleWindowsForSpace = 0.09900000ms
2018-08-15 18:37:33 DEBUG: chunkwm:Tiling:focused_desktop_mode
2018-08-15 18:37:33 DEBUG: chunkwm:Tiling:focused_window_float
2018-08-15 18:37:33 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayRemoved'
2018-08-15 18:37:33 DEBUG: 1104977153: display removed
2018-08-15 18:37:33 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayAdded'
2018-08-15 18:37:33 DEBUG: 69733568: display added
2018-08-15 18:37:33 DEBUG: chunkwm: processing event of type 'ChunkWM_WindowTitleChanged'
```

Note,  the display was never "resized" thus the fix in f604a5b5a6f072322c6436ce83cc626aca6ee17f does not work for this particular case. I believe in this log I both connected and disconnected the display back and forth.

In some other cases, I've also seen that the "DisplayMoved" event which also calls the `RecreateVirtualSpaceRegionsForDisplay` function. However in those cases it is called for the old display.

For example `459132913` is new display and `69733568` is old display that is being removed (seems moved event is triggered in some circumstances):

```
2018-08-15 19:26:52 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayAdded'
2018-08-15 19:26:52 DEBUG: 459132913: display added
2018-08-15 19:26:52 DEBUG: chunkwm: processing event of type 'ChunkWM_DisplayMoved'
2018-08-15 19:26:52 DEBUG: 69733568: display moved
#1423:plugin.mm:DisplayMovedHandler: assert failed 'DisplayRef'
fish: './bin/chunkwm' terminated by signal SIGSEGV (Address boundary error)
```

This particular assertion is not new from my changes though, I reproduce them on master too see #490. Since it is called for old display, this is likely why the assertion fails and why AXLibGetDisplayIdentifier can return null?

### CODE REVIEW

File: src/plugins/tiling/plugin.mm [hunk 1]
```diff
@@ -1450,11 +1451,6 @@ void CreateWindowTree()
          * in the current session.
          */
 
-        char *Ref = CopyCFStringToC(DisplayRef);
-        printf("display with ref '%s' connected!\n", Ref);
```
Comment (kylef) [inline src/plugins/tiling/plugin.mm]: I've removed this log as it wasn't being ran before, and other log messages are already fired for this event from the core callback handler.

**Comment by koekeishiya** [comment]
The situation is more complex than this though. See #40 and related issues.

**Comment by kylef** [comment]
@koekeishiya Yes, I understand the long-term plan is to place windows back to monitors they was originally present. This change is an improvement to the current behaviour in f604a5b5a6f072322c6436ce83cc626aca6ee17f which makes sure they fit on the new display. fixing #414. Is there any problems with this change that would cause other incorrect behaviour?

Currently when you follow the steps in my PR description the following happens:

#### External monitor view

screen shot 2018-08-16 at 09 53 39

#### Unplugged external monitor and then opened internal display

screen shot 2018-08-16 at 09 54 46

#### After patch (Unplugged external monitor and then opened internal display)

screen shot 2018-08-16 at 09 52 54

**Comment by koekeishiya** [comment]
Problem is that, even with this fix, the window is tiled on two desktops in chunkwm's representation, which can cause all weird crap to happen such as the window being moved to the other monitor when some interaction is going on over there.

**Comment by kylef** [comment]
> the window is tiled on two desktops in chunkwm's representation

Oh, I assumed the desktop on which the window was on is purely moved from the removed display to the new display in this case. RecreateVirtualSpaceRegionsForDisplay to refresh the internal state of the current (only) desktop.

**Comment by kylef** [comment]
> Problem is that, even with this fix, the window is tiled on two desktops in chunkwm's representation

@koekeishiya If the internal representation in chunkwm is wrong and we can't do anything about it. Perhaps chunkwm should exit under the cases where the internal state is wrong and cannot be resolved. That way at least launch managers and such can restart chunkwm upon its exit and user can continue.

What do you think?

**Comment by clevinson** [comment]
This is an issue for me as well. I'd love to know what people are doing as a scripted/temporary resolution for this if it is still unrselved.

**Pull request closed without merging**